### PR TITLE
Add mit-plv/bedrock2-ci to CI

### PR DESCRIPTION
**Pull request by andres-erbsen**
Previously discussed at https://github.com/coq/coq/pull/7878, https://github.com/mit-plv/bedrock2/issues/12, and https://github.com/mit-plv/bedrock2-ci/pull/1, development CI configuration at https://github.com/mit-plv/bedrock2-ci/blob/master/.travis.yml. I followed what Jason did to add cross-crypto at https://github.com/coq/coq/pull/7473.

### CODE REVIEW

File: .circleci/config.yml
Comment (SkySkimmer) [inline .circleci/config.yml]: We're not adding to circle any more.

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: This is wrong.
- not bignums
- the yaml variable is ci-template

File: dev/ci/ci-basic-overlay.sh
Comment (SkySkimmer) [inline dev/ci/ci-basic-overlay.sh]: `bignums` is wrong (bad copy paste?).

File: dev/ci/c-bedrock2.sh
Comment (SkySkimmer) [inline dev/ci/c-bedrock2.sh]: Why is this script talking about cross-crypto?

**Review by SkySkimmer — DISMISSED** [review]
Please fix the copy pasting errors.

**Comment by SkySkimmer** [comment]
Do we want to backport this? Is it compatible with Coq 8.8?

**Comment by andres-erbsen** [comment]
Fixed the noted copy-paste errors.

bedrock2 is built with 8.7 and 8.8 (and now master) so I think backporting would make sense.

File: dev/ci/c-bedrock2.sh
Comment (maximedenes) [inline dev/ci/c-bedrock2.sh]: This is a hack, can't we get upstream support for it instead?

File: dev/ci/c-bedrock2.sh
Comment (maximedenes) [inline dev/ci/c-bedrock2.sh]: Why not point to the bedrock2 repo instead? This will make it more difficult for us to submit patches upstream.

**Review by maximedenes — DISMISSED** [review]
If I understand correctly, this setup will require two patches each time we break compatibility (one for bedrock2, and one to bump the submodule in bedrock2-ci). Can we use the bedrock2 repo directly, instead?

File: dev/ci/c-bedrock2.sh
Comment (samuelgruetter) [reply dev/ci/c-bedrock2.sh]: I also think that this is a hack and I'll remove the `EXPECTED_COQC_VERSION` check from bedrock2.

File: dev/ci/c-bedrock2.sh
Comment (samuelgruetter) [reply dev/ci/c-bedrock2.sh]: I see, so you're saying that if a Coq change breaks bedrock2 and you create a patch, you'd like to make only one PR (against bedrock2) and avoid an extra PR against bedrock2-ci, right?

bedrock2 depends on two other repos (bbv and riscv-coq), and they have to be separate repos. We did spend some thought on how to put everything together for CI, see eg [here](https://github.com/mit-plv/bedrock2/issues/12#issuecomment-399248610). The main concerns are diamond dependencies and having a fast workflow.

So, an alternative CI setup which would be more convenient for the Coq developers would be to have bbv and riscv-coq as git submodules of bedrock2. Like this, if a patch for bbv or riscv-coq is needed, it would still require two PRs, but patches for bedrock2 (where most of the code lives) would only require one PR. This seems possible, but once we get bitten by diamond dependencies on bbv, we'll have to complicate the build setup of bedrock2 a bit by allowing it to depend on a bbv at a different path, ignoring its bbv submodule. I'll keep thinking about how to do that nicely.

@maximedenes  does that make sense, or did I misunderstand something?

**Comment by Zimmi48** [comment]
@andres-erbsen I don't see a problem with your own setup to run CI but I agree with @maximedenes that we shouldn't have to care about it. It seems that the problem here is similar to the one we have with stdpp-iris-lambda-rust and which was handled in a (different) ad hoc way. I think the problem is on our side: we should support such standard use case and have a generic way to answer the questions you were debating in mit-plv/bedrock2#12. I'll try to open an issue about this soon to have a generic discussion.

**Review by SkySkimmer — APPROVED** [review]
You need to squash commits.

File: .gitlab-ci.yml
Comment (ejgallego) [inline .gitlab-ci.yml]: Alpha order is incorrect here.

File: .travis.yml
Comment (ejgallego) [inline .travis.yml]: Don't update Travis anymore.

File: Makefile.ci
Comment (ejgallego) [inline Makefile.ci]: Problem with alpha sorting.

**Review by ejgallego — CHANGES_REQUESTED** [review]

**Review by ejgallego — APPROVED** [review]
I am personally OK adding this as-is and improving the setup later, however I shall respect @maximedenes choice in that regard, so won't merge until I get green light from him.

**Comment by maximedenes** [comment]
> I am personally OK adding this as-is and improving the setup later, however I shall respect @maximedenes choice in that regard, so won't merge until I get green light from him.

I was about to ask your opinion and ready to follow it :)

**Comment by ejgallego** [comment]
> I was about to ask your opinion and ready to follow it :)

My opinion is that the testing benefit may outweight the temporary fixing annoyances; in particular as I hope we don't break bedrock so often; but I am clearly biased towards more extensive testing.

**Comment by maximedenes** [comment]
We can merge as is and when we submit a bedrock2 patch, ask upstream to update the submodule.

**Comment by samuelgruetter** [comment]
Not sure if everyone noticed but about 9 hours ago, I changed the CI setup of bedrock2, see [this comment](https://github.com/mit-plv/bedrock2/issues/12#issuecomment-400571524).
The dependencies (bbv and riscv-coq) are now submodules of bedrock2, and the "umbrella repo" bedrock2-ci does not appear in this PR any more.
Does that address your patch workflow concerns?

**Comment by ejgallego** [comment]
> Does that address your patch workflow concerns?

Hi @samuelgruetter I am fine with the setup for now, mergin as soon as CI passes.

**Comment by ejgallego** [comment]
@andres-erbsen note the silly linter error in Travis, can you fix and push again? Thanks.

**Comment by ejgallego** [comment]
While you are at it with the linter please rebase to master.

**Pull request merged**
